### PR TITLE
User-level statically defined tracing (USDT)

### PR DESCRIPTION
**Pull request by toniSta**
This pr introduces USDT.

- both Dtrace and SystemTap scripts for performance analysis
- conversion of SystemTap-style probes to Dtrace-style probes
- Script, that analysis time consumed per operator within a SQL query

Some documentation can be found in this [blogpost](https://medium.com/hyrise/having-a-closer-look-under-the-hood-3f186ba78bdd) or in the [wiki](https://github.com/hyrise/hyrise/wiki/Profiling#usdt-with-systemtap-and-dtrace).

### CODE REVIEW

File: tracer_scripts/job_start_end.d [hunk 1]
```diff
@@ -0,0 +1,44 @@
+#!/usr/sbin/dtrace -s
```
Comment (mrks) [inline tracer_scripts/job_start_end.d]: Is this the same as `jobs_start_end.stp`? If yes, please fix the naming and add a short comment.

File: src/lib/utils/tracing/systemtap.hpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+// In order to be compatible with Mac dtrace probes, both provider and probe name have to be uppercase
+constexpr bool is_valid_name(const char* name) {
+  do {
+    if (*name >= 'a' && *name <= 'z') return false;
+    if (*name == '-') return false;
+  } while (*++name != '\0');
+  return true;
+}
+
+#if defined(__APPLE__) || defined(__MACOS__)
```
Comment (mrks) [inline src/lib/utils/tracing/systemtap.hpp]: Why do you check for apple in a file called "systemtap"? Maybe replace with a generic name?

File: src/lib/utils/tracing/probe_definitions.hpp [hunk 3]
```diff
@@ -0,0 +1,39 @@
+#ifndef HYRISE_TASKS_PER_STATEMENT
```
[the remark this answers is not in the document]
Comment (toniSta) [reply src/lib/utils/tracing/probe_definitions.hpp]: This file is generated automatically and thus, is not supposed to be commited at all.

File: scripts/track_time_by_operator.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env python
```
[the remark this answers is not in the document]
Comment (toniSta) [reply scripts/track_time_by_operator.py]: When starting a SystemTap session you can specify a binary (plus parameters) which will be executed in a new shell.

[the remark this answers is not in the document]
Comment (toniSta) [reply scripts/track_time_by_operator.py]: Basically, this script can be applied to each systemtap script. However, I wanted to provide it for an easy analysis of the time consumption per operator.

[the remark this answers is not in the document]
Comment (mrks) [reply scripts/track_time_by_operator.py]: I still fail to see the purpose of this script. Doesn't it simply call `sudo stap tracer_scripts/query_analyzer.stp -DMAXMAPENTRIES=10000 -DMAXACTION=100000000 -G debug=0 -c ...`?

If yes, wouldn't it be sufficient to add this command line as a comment to the stp file?

[the remark this answers is not in the document]
Comment (toniSta) [reply scripts/track_time_by_operator.py]: Yes, you are right. I will delete this file and add a comment to the stp file.

File: src/lib/CMakeLists.txt [hunk 5]
```diff
@@ -581,6 +583,23 @@ if (${ENABLE_NUMA_SUPPORT})
     set(LIBRARIES ${LIBRARIES} ${NUMA_LIBRARY} hpinuma_msource_s)
 endif()
 
+# Generate a macro file containing probe definitions for dtrace users. As soon as a probe is not present
+# in `provider.h` a compile error would be thrown. With this ugly solution we can avoid this.
+# First we find all probes within the code (e.g. DTRACE_PROBE1(HYRISE, TEST, arg1))
+# Than the regex greps provider and probe name and concactenates them with '_' (e.g. HYRISE_TEST)
+# The probes are than written into a file in order to define the probes
```
Comment (mrks) [inline src/lib/CMakeLists.txt]: typo: then

File: .gitignore [hunk 6]
```diff
@@ -22,6 +22,8 @@ lqp*.png
 compile_times.txt
 exported_table_statistics.json
 version.hpp
+provider.h
+probe_definitions.hpp
```
Comment (mrks) [inline .gitignore]: Please add path

File: src/lib/CMakeLists.txt (hunk 5 above)
Comment (mrks) [reply src/lib/CMakeLists.txt]: Line above too

File: tracer_scripts/query_execution_time.d [hunk 7]
```diff
@@ -0,0 +1,13 @@
+#!/usr/sbin/dtrace -s
```
Comment (mrks) [inline tracer_scripts/query_execution_time.d]: How do I run this?
```
[:~/Projekte/Opossum/Git] mp18/usdt+ ± sudo ./tracer_scripts/query_execution_time.d -c ./build/hyriseConsole
dtrace: system integrity protection is on, some features will not be available

[PERF] Hyrise is running as a debug build. at src/lib/utils/performance_warning.cpp:9
	Performance can be affected. This warning is only shown once.

dtrace: failed to compile script ./tracer_scripts/query_execution_time.d: line 3: probe description hyrise34815:::query_start does not match any probes
[:~/Projekte/Opossum/Git] mp18/usdt+ ± grep -r query_start src
src/lib/utils/tracing/provider.d:        probe query_start(char* query, int task_size, uintptr_t this_pointer);
```

File: src/lib/CMakeLists.txt [hunk 8]
```diff
@@ -581,6 +583,23 @@ if (${ENABLE_NUMA_SUPPORT})
...
+
+set(PROBES_FILE "${CMAKE_SOURCE_DIR}/src/lib/utils/tracing/probe_definitions.hpp")
+set(LINE "#ifndef &\\n#define &(...)\\n#endif\\n")
+add_custom_command (
+    OUTPUT ${PROBES_FILE}
+    DEPENDS ${SOURCES}
+    COMMAND grep -r --include=*.[ch]pp --exclude=probes.hpp --exclude=probe_definitions.hpp -h DTRACE_PROBE ${CMAKE_SOURCE_DIR} |  sed "s/[^(]*(\\([^,]*\\), \\([^(,|))]*\\).*/\\1_\\2/" | uniq | xargs -n 1 -I & printf "${LINE}" > ${PROBES_FILE}
```
Comment (mrks) [inline src/lib/CMakeLists.txt]: This also includes probes that are commented out. Also, please add a comment to the regexp.

File: tracer_scripts/job_start_end.d (hunk 1 above)
Comment (mrks) [reply tracer_scripts/job_start_end.d]: ping

Comment (toniSta) [reply tracer_scripts/job_start_end.d]: Yes, it is the same as `jobs_start_end.stp`. I fixed the naming and added a comment.

File: tracer_scripts/query_execution_time.d (hunk 7 above)
Comment (toniSta) [reply tracer_scripts/query_execution_time.d]: System integrity protection is enabled on your device. This prohibits the usage of USDT with dtrace. I added a paragraph to the [wiki](https://github.com/hyrise/hyrise/wiki/Profiling#mac-os). Look at this [post](https://stackoverflow.com/a/33584192) to solve this issue.

File: src/CMakeLists.txt [hunk 9]
```diff
@@ -60,7 +60,7 @@ endif()
 add_definitions("-DSOURCE_PATH_SIZE=${SOURCE_PATH_SIZE}")
 
 # Global flags and include directories
-add_compile_options(-std=c++17 -pthread -Wall -Wextra -pedantic -Werror -Wno-unused-parameter)
+add_compile_options(-std=c++17 -pthread -Wall -Wextra -pedantic -Werror -Wno-unused-parameter -Wno-dollar-in-identifier-extension)
```
Comment (mrzzzrm) [inline src/CMakeLists.txt]: Please add a short comment why this (rather obscure?) flag is necessary

**Pull request merged**